src/shapes/bolt.1
.TH BOLT 1 BRL-CAD
.\"                         B O L T . 1
.\" BRL-CAD
.\"
.\" Copyright (c) 2005-2009 United States Government as represented by
.\" the U.S. Army Research Laboratory.
.\"
.\" Redistribution and use in source (Docbook format) and 'compiled'
.\" forms (PDF, PostScript, HTML, RTF, etc), with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code (Docbook format) must retain the
.\" above copyright notice, this list of conditions and the following
.\" disclaimer.
.\"
.\" 2. Redistributions in compiled form (transformed to other DTDs,
.\" converted to PDF, PostScript, HTML, RTF, and other formats) must
.\" reproduce the above copyright notice, this list of conditions and
.\" the following disclaimer in the documentation and/or other
.\" materials provided with the distribution.
.\"
.\" 3. The name of the author may not be used to endorse or promote
.\" products derived from this documentation without specific prior
.\" written permission.
.\"
.\" THIS DOCUMENTATION IS PROVIDED BY THE AUTHOR AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
.\" OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS DOCUMENTATION, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\".\".\"
.SH NAME
bolt \- creates an mged file containing bolts
.SH SYNOPSIS
.B bolt
[ options... ]
.SH DESCRIPTION
bolt is a program to create an mged data base of one of four
different kinds of bolts.  Up to twenty-six bolts of one type
and size may
be created.  bolt uses libwdb to create an mged data base.  This
program may be run interactively or the user may specify options on
a command line.  If the user chooses to run the program interactively
he answers the questions as the program prompts him.  Below are the
options that can be used on the command line.
.SS Options
.TP "\w'-G ``n cflag gflag vsize\'\'\     |'u"
.BI \-o#\^
Type of bolt to be created:  1=>bolt head; 2=>bolt head and washer;
3=>bolt head, washer, and bolt stem; and 4=>bolt head and bolt stem.
.TP
.BI \-fname.g\^
Mged file name.
.TP
.BI \-n#\^
The number of bolts to be created.  This number must be less than or
equal to 26 (or it will be set to 26).
.TP
.BI \-hd#\^
Diameter of bolt head flat edge to flat edge in millimeters.
.TP
.BI \-hh#\^
Height of bolt head in millimeters.
.TP
.BI \-wd#\^
Diameter of washer in millimeters.
.TP
.BI \-wh#\^
Height of washer in millimeters.
.TP
.BI \-sd#\^
Diameter of bolt stem in millimeters.
.TP
.BI \-sh#\^
Height of bolt stem in millimeters.
.SH EXAMPLE
The following is an example from an interactive session.
.sp
.nf
$ bolt
Enter option:
	1 - bolt head
	2 - bolt head & washer
	3 - bolt head, washer, & stem
	4 - bolt head & stem
1
Enter name of mged file to be created (25 char max).
	bolt.g
Enter the number of bolts to be created (26 max).
	3
Enter diameter (flat edge to flat edge) & height of bolt head.
	30 10

option:  1 - bolt head
file:  bolt.g
head diameter:  30.000000, & height:  10.000000
washer diameter:  0.000000, & height:  0.000000
stem diameter:  0.000000, & height:  0.000000
number of bolts:  3

.fi
The following is the same example where all dimensions are on the
command line.
.nf

	bolt -o1 -fbolt.g -n3 -hd30 -hh10

.fi
In both these examples three bolt heads (option 1) are created with
a diameter of 30mm and a height of 10mm.  The mged file created in
each case is bolt.g.

.SH SEE ALSO
handle(1), window(1), winfrm(1), gastank(1)

.SH AUTHOR
Susan A. Coates


NEWS
						 -*- coding: utf-8 -*-
BRL-CAD Release Notes
=====================

This document provides a transcript of release notes and user-visible
changes for each release of BRL-CAD.  The individual changes are
listed in chronological order with the most recent changes listed
first and are grouped by release.  Each release also includes optional
descriptive text that emphasizes or further describes significant
changes made.  See document footer for additional details.


----------------------------------------------------------------------
--- 2008-XX-XX  Release 7.XX.X                                     ---
----------------------------------------------------------------------

* mged's inside command no longer uses the geometry currently being
	edited for the outside solid - Bob Parker
* reported and fixed mged documentation eto typein bug - Robert Penny
* fixed mged itcl.tcl startup failure prior to install - Sean Morrison
* fixed bot_dump -u command line option for setting units - Bob Parker
* changed automatic 'make' sizes for various primitives - Bob Parker
* fixed bug with mged 'make' command automatic keypoints - Bob Parker
* renamed all mk_ tools to remove the mk_ prefix - Sean Morrison
* more conversion of documentation to Docbook format - Janine Gettier
* removed the unused 'k'eypoint point on extrudes - Sean Morrison
* added new dem-g terrain importer - Richard Weiss
* layout tweaks to display of selection lists in archer - Doug Howard
* activate primitive creation buttons in archer - Bob Parker
* activate wizard plugins and add tire wizard to archer - Bob Parker
* new initial manual page for bot_dump - Sean Morrison
* added editing support for hyperboloid primitive to mged - Cliff Yapp
* new 'lens' procedural tool for making optical lenses - Cliff Yapp


----------------------------------------------------------------------
--- 2008-11-04  Release 7.14.0                                     ---
----------------------------------------------------------------------

Of the many enhancements included in this release, MGED gets a
powerful new command called 'search' that is very similar to the UNIX
'find' command but for geometry hierarchies.  Included in the
implementation of 'search' is the ability to list geometry that
matches a specified name using globbing patterns or regular
expressions.  You can also search for objects that have attributes
set, that are referenced in combinations, are at a particular depth in
the geometry hierarchy, and much more.  You can combine search
expressions together using boolean logic as well for even more
powerful operations.

* remove -empty option from mged "search" command - Cliff Yapp
* add -nnodes node count option to mged "search" command - Cliff Yapp
* upgraded libpng to 1.2.32 - Erik Greenwald
* fixed memory leak with in-memory geometry databases - Daniel Roßberg
* mged help menu is restructured for improved usability - Cliff Yapp
* update openNURBS to version 4.0.2007.0718 - Cliff Yapp
* handle (i.e., ignore) comments in asc2g conversions - Erik Greenwald
* fix crash in iges-g when run with -n option - John Anderson
* enhanced documentation management system within mged - Cliff Yapp
* conversion of documentation to Docbook format, editing, and review
	- Cliff Yapp, Janine Gettier
* improved benchmark suite usage and instructions - Sean Morrison
* improved mesh normals from Pro/E exporter - Bob Parker
* fixed 'bspline' surface primitive wireframe plot bug - John Anderson
* removed bot-raw, functionality replaced by bot_dump - Sean Morrison
* abort during non-planar collapse or simplify of NMG  - John Anderson
* jove is deprecated and will no longer be maintained - Sean Morrison
* plugged attribute memory leak in all database tools - Daniel Roßberg
* DSP tessellation honors cut type and handles zeros - John Anderson
* fixed 'gets' command input bug in mged - Sean Morrison, Bob Parker
* improved fast4-g DOS file line-ending support - Sean Morrison
* numerous documentation fixes to manual pages - Cliff Yapp
* bot_dump exporter for extracting BoT meshes unevaluated - Bob Parker
* fixed bug parsing long command lines in mged - Sean Morrison
* tessellation support for the DSP primitive - John Anderson
* additional logging options in Pro/E exporter - Bob Parker
* Pro/E exporter enhancements for dynamic tess tolerance - Bob Parker
* new search command enabling Unix find style searches - C. Yapp
* mged no longer requires PATH for internal commands - Sean Morrison
* mged 'edcolor' command turns into 'color -e' - Sean Morrison
* nirt dest command now avoids overwriting existing files - Cliff Yapp
* fixed mged_players error on tra in classic mode mged  - Bob Parker
* fix attr behavior when setting the region attribute - C. Yapp
* identified fix for "Illegal instruction" bug in rt - Simon Clubley
* mged get_regions and reid ignore duplicate regions - Sean Morrison
* added procedural geometry tools to the Windows port
	- Bob Parker, Andre Castelo
* integrated tgf-g with the default compile/install - Daniel Roßberg
* deprecated the mged '-n' option, use '-c' instead - Sean Morrison
* added the ability to invoke mged through icons/menus - Sean Morrison
* new -a option to mged 'killtree' for killing references - Bob Parker
* added new 'killrefs' command to mged - Bob Parker
* added -C color option to the mged 'E' command - Bob Parker
* fixed curvature calc bug in ehy primitive - Timothy Van Ruitenbeek
* added compilation support for MSVC 9 - André Castelo


----------------------------------------------------------------------
--- 2008-08-19  Release 7.12.6                                     ---
----------------------------------------------------------------------

This release enhances the formatting options available "out of the
box" for the NIRT tool, allowing users to easily select different
data reporting options without having to define their own custom
configuration files.  CSV output options also allow for quick

src/shapes/bolt.c
/*                          B O L T . C
 * BRL-CAD
 *
 * Copyright (c) 2004-2008 United States Government as represented by
 * the U.S. Army Research Laboratory.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this file; see the file named COPYING for more
 * information.
 */
/** @file bolt.c
 *
 * Program to make a bolt using libwdb.  The objects will be in mm.
 *
 */

#include "common.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "vmath.h"
#include "db.h"
#include "raytrace.h"
#include "wdb.h"


int
main(int argc, char **argv)
{
    struct rt_wdb *fpw;		/* File to be written to. */
    char filemged[26];		/* Mged file create. */
    double hd, hh;		/* Diameter & height of bolt head. */
    double wd, wh;		/* Diameter & height of washer. */
    double sd, sh;		/* Diameter & height of bolt stem. */
    double leg, hyp;		/* Length of leg & hypotenus of triangle. */
    point_t pts[8];		/* Eight points of arb8. */
    point_t bs;			/* Base of rcc. */
    vect_t ht;			/* Height of rcc. */
    fastf_t rad;		/* Radius of rcc. */
    int iopt;			/* Type of bolt to be built: 1=>bolt head,  */
				/*  2=>head & washer; 3=>head, washer, &  */
				/* stem; 4=>head & stem. */
    int i, j, k;		/* Loop counters. */
    char *temp;			/* Temporary character string. */
    char temp1[16];		/* Temporary character string. */

    char solnam[9];		/* Solid name. */
    char regnam[9];		/* Region name. */
    char grpnam[9];		/* Group name. */
    int numblt;			/* Number of bolts to be created (<=26). */

    struct wmember comb;	/* Used to make regions. */
    struct wmember comb1;	/* Used to make groups. */

    /* Zero all dimnsions of bolt. */
    iopt = 0;
    hd = 0;
    hh = 0;
    wd = 0;
    wh = 0;
    sd = 0;
    sh = 0;

    /*   Set up solid, region, and group names. */
    solnam[0] = 's';
    solnam[1] = '.';
    solnam[2] = 'b';
    solnam[3] = 'o';
    solnam[4] = 'l';
    solnam[5] = 't';
    solnam[6] = ' ';
    solnam[7] = '#';
    solnam[8] = '\0';
    regnam[0] = 'r';
    regnam[1] = '.';
    regnam[2] = 'b';
    regnam[3] = 'o';
    regnam[4] = 'l';
    regnam[5] = 't';
    regnam[6] = ' ';
    regnam[7] = '#';
    regnam[8] = '\0';
    grpnam[0] = 'b';
    grpnam[1] = 'o';
    grpnam[2] = 'l';
    grpnam[3] = 't';
    grpnam[4] = ' ';
    grpnam[5] = '\0';

    /* If there are no arguments ask questions. */
    if (argc == 1)
    {
	/* START # 1  */

	/* Find type of bolt to build. */
	(void)printf("Enter option:\n");
	(void)printf("\t1 - bolt head\n");
	(void)printf("\t2 - bolt head & washer\n");
	(void)printf("\t3 - bolt head, washer, & stem\n");
	(void)printf("\t4 - bolt head & stem\n");
	(void)fflush(stdout);
	(void)scanf("%d", &iopt);

	/* Get file name of mged file to be created. */
	(void)printf("Enter name of mged file to be created (25 char max).\n\t");
	(void)fflush(stdout);
	(void)scanf("%26s", filemged);

	/* Find the number of bolts to be created (<=26). */
	(void)printf("Enter the number of bolts to be created (26 max).\n\t");
	(void)fflush(stdout);
	(void)scanf("%d", &numblt);
	if (numblt > 26) numblt = 26;

	/* Find dimensions of the bolt. */
	/* Find dimensions of head first. */
	(void)printf("Enter diameter (flat edge to flat edge) & height of ");
	(void)printf("bolt head.\n\t");
	(void)fflush(stdout);
	(void)scanf("%lf %lf", &hd, &hh);
	/* Find dimensions of washer if necessary. */
	if ((iopt == 2) || (iopt == 3))
	{
	    (void)printf("Enter diameter & height of washer.\n\t");
	    (void)fflush(stdout);
	    (void)scanf("%lf %lf", &wd, &wh);
	}
	/* Find dimensions of bolt stem if necessary. */
	if ((iopt == 3) || (iopt == 4))
	{
	    (void)printf("Enter diameter & height of bolt stem.\n\t");
	    (void)fflush(stdout);
	    (void)scanf("%lf %lf", &sd, &sh);
	}

    }							/* END # 1  */

    /* If there are arguments do not ask any questions.  Get the */
    /* answers from the arguments. */
    else
    {
	/* START # 2  */

	/* List of options. */
	/*	-o# - # = 1 => bolt head */
	/*	          2 => head & washer */
	/*	          3 => head, washer, & stem */
	/*	          4 => head & stem */
	/*	-fname - name = name of .g file */
	/*  	-n# - # = number of bolts to create (<=26)  */
	/*	-hd# - # = head diameter */
	/*	-hh# - # = head height */
	/*	-wd# - # = washer diameter */
	/*	-wh# - # = washer height */
	/*	-sd# - # = stem diameter */
	/*	-sh# - # = stem height */

	for (i=1; i<argc; i++)
	{
	    /* START # 3  */
	    /* Put argument into temporary character string. */
	    temp = argv[i];

	    /*  -o - set type of bolt to make. */
	    if (temp[1] == 'o')
	    {
		/* START # 4  */
		if (temp[2] == '1') iopt = 1;
		if (temp[2] == '2') iopt = 2;
		if (temp[2] == '3') iopt = 3;
		if (temp[2] == '4') iopt = 4;

	    }						/* END # 4  */

	    /*  -f - mged file name. */
	    else if (temp[1] == 'f')
	    {
		/* START # 5  */
		j = 2;
		k = 0;
		while ((temp[j] != '\0') && (k < 25))
		{
		    /* START # 6  */
		    filemged[k] = temp[j];
		    j++;
		    k++;
		}					/* END # 6  */
		filemged[k] = '\0';
	    }						/* END # 5  */

	    /*  -n - number of bolts to be created. */
	    else if (temp[1] == 'n')
	    {
		/* START # 6.05  */
		/* Set up temporary character string, temp1. */
		j = 2;
		k = 0;
		while ((temp[j] != '\0') && (k < 15))
		{
		    temp1[k] = temp[j];
		    j++;
		    k++;
		}
		temp1[k] = '\0';
		(void)sscanf(temp1, "%d", &numblt);
		if (numblt > 26) numblt = 26;
	    }						/* END # 6.05  */

	    /* Take care of all other arguments. */
	    else
	    {
		/* START # 6.1  */
		/* Set temporary character string, temp1. */
		j = 3;
		k = 0;
		while ((temp[j] != '\0') && (k < 15))
		{
		    temp1[k] = temp[j];
		    j++;
		    k++;
		}
		temp1[k] = '\0';

		/*  -hd & -hh - head diameter & height. */
		if (temp[1] == 'h')
		{
		    /* START # 7  */
		    if (temp[2] == 'd')	/* Head diameter. */
		    {
			(void)sscanf(temp1, "%lf", &hd);
		    }
		    else if (temp[2] =='h')	/* Head height. */
		    {
			(void)sscanf(temp1, "%lf", &hh);
		    }
		}					/* END # 7  */

		/*  -wd & -wh - washer diameter & height. */
		else if (temp[1] == 'w')
		{
		    /* START # 8  */
		    if (temp[2] == 'd')	/* Washer diameter. */
		    {
			(void)sscanf(temp1, "%lf", &wd);
		    }
		    else if (temp[2] == 'h')	/* Washer height. */
		    {
			(void)sscanf(temp1, "%lf", &wh);
		    }
		}					/* END # 8  */

		/*  -sd & -sh - stem washer diameter & height. */
		else if (temp[1] == 's')
		{
		    /* START # 9  */
		    if (temp[2] == 'd')	/* Stem diameter. */
		    {
			(void)sscanf(temp1, "%lf", &sd);
		    }
		    else if (temp[2] == 'h')	/* Stem height. */
		    {
			(void)sscanf(temp1, "%lf", &sh);
		    }
		}					/* END # 9  */
	    }						/* END # 6.1  */

	}						/* END # 3  */
    }							/* END # 2  */

    /* Print out bolt dimensions. */
    (void)printf("\noption:  %d - ", iopt);
    if (iopt == 1) (void)printf("bolt head\n");
    if (iopt == 2) (void)printf("head & washer\n");
    if (iopt == 3) (void)printf("head, washer, & stem\n");
    if (iopt == 4) (void)printf("head & stem\n");
    (void)printf(".g file:  %s\n", filemged);
    (void)printf("head diameter:  %f, & height:  %f\n", hd, hh);
    (void)printf("washer diameter:  %f, & height:  %f\n", wd, wh);
    (void)printf("stem diameter:  %f, & height:  %f\n", sd, sh);
    (void)printf("number of bolts:  %d\n\n", numblt);
    (void)fflush(stdout);

    /* Open mged file for writing to. */
    fpw = wdb_fopen(filemged);

    /* Write ident record. */
    mk_id(fpw, "bolts");

    for (i=0; i<numblt; i++)	/* Loop for each bolt created. */
    {
	/* START # 20  */

	/* Create all solids needed. */
	/* Create solids of bolt head. */
	leg = tan(M_PI / 6.) * hd / 2.;
	hyp = leg * leg + (hd / 2.) * (hd / 2.);
	hyp = sqrt(hyp);
	/* Bolt head is two solids, create first solid. */
	pts[0][0] = (fastf_t) ((-hd) / 2.);
	pts[0][1] = (fastf_t)leg;
	pts[0][2] = (fastf_t)hh;
	pts[1][0] = (fastf_t)0.;
	pts[1][1] = (fastf_t)hyp;
	pts[1][2] = (fastf_t)hh;
	pts[2][0] = (fastf_t)0.;
	pts[2][1] = (fastf_t)(-hyp);
	pts[2][2] = (fastf_t)hh;
	pts[3][0] = (fastf_t) ((-hd) / 2.);
	pts[3][1] = (fastf_t)(-leg);
	pts[3][2] = (fastf_t)hh;
	pts[4][0] = (fastf_t) ((-hd) / 2.);
	pts[4][1] = (fastf_t)leg;
	pts[4][2] = (fastf_t)0.;
	pts[5][0] = (fastf_t)0.;
	pts[5][1] = (fastf_t)hyp;
	pts[5][2] = (fastf_t)0.;
	pts[6][0] = (fastf_t)0.;
	pts[6][1] = (fastf_t)(-hyp);
	pts[6][2] = (fastf_t)0.;
	pts[7][0] = (fastf_t) ((-hd) / 2.);
	pts[7][1] = (fastf_t)(-leg);
	pts[7][2] = (fastf_t)0.;
	solnam[6] = 97 + i;
	solnam[7] = '1';
	mk_arb8(fpw, solnam, &pts[0][X]);

	/* Create second solid. */
	pts[0][0] = (fastf_t) (hd / 2.);
	pts[0][1] = (fastf_t)leg;
	pts[0][2] = (fastf_t)hh;
	pts[1][0] = (fastf_t)0.;
	pts[1][1] = (fastf_t)hyp;
	pts[1][2] = (fastf_t)hh;
	pts[2][0] = (fastf_t)0.;
	pts[2][1] = (fastf_t)(-hyp);
	pts[2][2] = (fastf_t)hh;
	pts[3][0] = (fastf_t) (hd / 2.);
	pts[3][1] = (fastf_t)(-leg);
	pts[3][2] = (fastf_t)hh;
	pts[4][0] = (fastf_t) (hd / 2.);
	pts[4][1] = (fastf_t)leg;
	pts[4][2] = (fastf_t)0.;
	pts[5][0] = (fastf_t)0.;
	pts[5][1] = (fastf_t)hyp;
	pts[5][2] = (fastf_t)0.;
	pts[6][0] = (fastf_t)0.;
	pts[6][1] = (fastf_t)(-hyp);
	pts[6][2] = (fastf_t)0.;
	pts[7][0] = (fastf_t) (hd / 2.);
	pts[7][1] = (fastf_t)(-leg);
	pts[7][2] = (fastf_t)0.;
	solnam[7] = '2';
	mk_arb8(fpw, solnam, &pts[0][X]);

	/* Create washer if necessary. */
	if ((iopt == 2) || (iopt == 3))
	{
	    bs[0] = (fastf_t)0.;
	    bs[1] = (fastf_t)0.;
	    bs[2] = (fastf_t)0.;
	    ht[0] = (fastf_t)0.;
	    ht[1] = (fastf_t)0.;
	    ht[2] = (fastf_t)(-wh);
	    rad = (fastf_t) (wd / 2.);
	    solnam[7] = '3';
	    mk_rcc(fpw, solnam, bs, ht, rad);
	}

	/* Create bolt stem if necessary. */
	if ((iopt == 3) || (iopt == 4))
	{
	    bs[0] = (fastf_t)0.;
	    bs[1] = (fastf_t)0.;
	    bs[2] = (fastf_t)0.;
	    ht[0] = (fastf_t)0.;
	    ht[1] = (fastf_t)0.;
	    ht[2] = (fastf_t)(-sh);
	    rad = (fastf_t) (sd / 2.);
	    solnam[7] = '4';
	    mk_rcc(fpw, solnam, bs, ht, rad);
	}

	/* Create all regions. */
	/* First must initialize list. */
	BU_LIST_INIT(&comb.l);

	/* Create region for first half of bolt head. */
	solnam[7] = '1';
	(void)mk_addmember(solnam, &comb.l, NULL, WMOP_INTERSECT);
	solnam[7] = '2';
	(void)mk_addmember(solnam, &comb.l, NULL, WMOP_SUBTRACT);
	/* Subtract washer if it exists. */
	if ((iopt == 2) || (iopt == 3))
	{
	    solnam[7] = '3';
	    (void)mk_addmember(solnam, &comb.l, NULL, WMOP_SUBTRACT);
	}
	/* Subtract stem if it exists. */
	if ((iopt == 3) || (iopt == 4))
	{
	    solnam[7] = '3';
	    (void)mk_addmember(solnam, &comb.l, NULL, WMOP_SUBTRACT);
	}
	regnam[6] = 97 + i;
	regnam[7] = '1';
	mk_lfcomb(fpw, regnam, &comb, 1);

	/* Create region for second half of bolt head. */
	solnam[7] = '2';
	(void)mk_addmember(solnam, &comb.l, NULL, WMOP_INTERSECT);
	/* Subtract washer if it exists. */
	if ((iopt == 2) || (iopt == 3))
	{
	    solnam[7] = '3';
	    (void)mk_addmember(solnam, &comb.l, NULL, WMOP_SUBTRACT);
	}
	/* Subtract stem if it exists. */
	if ((iopt == 3) || (iopt == 4))
	{
	    solnam[7] = '4';
	    (void)mk_addmember(solnam, &comb.l, NULL, WMOP_SUBTRACT);
	}
	regnam[7] = '2';
	mk_lfcomb(fpw, regnam, &comb, 1);

	/* Create region for washer if it exists. */
	if ((iopt == 2) || (iopt == 3))
	{
	    solnam[7] = '3';
	    (void)mk_addmember(solnam, &comb.l, NULL, WMOP_INTERSECT);
	    /* Subtract bolt stem if it exists. */
	    if (iopt == 3)
	    {
		solnam[7] = '4';
		(void)mk_addmember(solnam, &comb.l, NULL, WMOP_SUBTRACT);
	    }
	    regnam[7] = '3';
	    mk_lfcomb(fpw, regnam, &comb, 1);
	}

	/* Create region for bolt stem if it exists. */
	if ((iopt == 3) || (iopt == 4))
	{
	    solnam[7] = '4';
	    (void)mk_addmember(solnam, &comb.l, NULL, WMOP_INTERSECT);
	    regnam[7] = '4';
	    mk_lfcomb(fpw, regnam, &comb, 1);
	}

	/* Create a group containing all bolt regions. */
	/* First initialize the list. */
	BU_LIST_INIT(&comb1.l);
	/* Add both bolt head regions to the list. */
	regnam[7] = '1';
	(void)mk_addmember(regnam, &comb1.l, NULL, WMOP_UNION);
	regnam[7] = '2';
	(void)mk_addmember(regnam, &comb1.l, NULL, WMOP_UNION);
	/* Add washer region if necessary. */
	if ((iopt == 2) || (iopt == 3))
	{
	    regnam[7] = '3';
	    (void)mk_addmember(regnam, &comb1.l, NULL, WMOP_UNION);
	}
	/* Add bolt stem region if necessary. */
	if ((iopt == 3) || (iopt == 4))
	{
	    regnam[7] = '4';
	    (void)mk_addmember(regnam, &comb1.l, NULL, WMOP_UNION);
	}
	/* Actually create the group. */
	grpnam[4] = 97 + i;
	mk_lfcomb(fpw, grpnam, &comb1, 0);

    }							/* END # 20  */

    /* Close mged file. */
    wdb_close(fpw);
    return 0;
}

/*
 * Local Variables:
 * mode: C
 * tab-width: 8
 * indent-tabs-mode: t
 * c-file-style: "stroustrup"
 * End:
 * ex: shiftwidth=4 tabstop=8
 */
